### PR TITLE
MM-18436 Improve perceived performance of emoji picker

### PR DESCRIPTION
**Pull request by sudheerDev**
#### Summary
  * Render only first two categories i.e recent and people on mount so we decrease the first meaningful paint time.
  * Render the rest of the categories on mount
  * PR also remove the title and category section of recent emojis
    when then do not exist

This effectively gets down first pain to under 200ms even with custom emojis when compared to 1.2s on daily as of now.

Part of apps team Seoul hackathon along with https://github.com/mattermost/mattermost-webapp/pull/3580

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-18436

### CODE REVIEW

File: components/emoji_picker/emoji_picker.test.jsx [hunk 1]
```diff
@@ -20,4 +39,119 @@ describe('components/emoji_picker/EmojiPicker', () => {
...
+    test('init snapshot', () => {
+        const wrapper = shallow(
+            <EmojiPicker {...baseProps}/>
+        );
+
+        wrapper.instance().emojiPickerContainer = {
+            offsetHeight: 200,
+        };
+
+        expect(wrapper).toMatchSnapshot();
```
Comment (sudheerDev) [inline components/emoji_picker/emoji_picker.test.jsx]: i don't think this snapshot adds much value and probably just noise. Just added incase. Let me know other feel the same.

File: components/emoji_picker/emoji_picker.jsx [hunk 2]
```diff
@@ -178,15 +187,24 @@ export default class EmojiPicker extends React.PureComponent {
...
         });
-        this.divHeight = this.emojiPickerContainer.offsetHeight;
+
+        if (this.emojiPickerContainer) {
+            this.divHeight = this.emojiPickerContainer.offsetHeight;
+        }
     }
 
     UNSAFE_componentWillUpdate(nextProps, nextState) { // eslint-disable-line camelcase
+        if (this.state.categories !== nextState.categories) {
```
Comment (enahum) [inline components/emoji_picker/emoji_picker.jsx]: nothing to do now on this PR but we should consider moving this to `componentDidUpdate`

File: components/emoji_picker/emoji_picker.jsx [hunk 3]
```diff
@@ -217,6 +235,23 @@ export default class EmojiPicker extends React.PureComponent {
         }
     }
 
+    componentWillUnmount() {
+        if (this.renderAllCategoriesFrame) {
+            window.cancelAnimationFrame(this.renderAllCategoriesFrame);
```
Comment (enahum) [inline components/emoji_picker/emoji_picker.jsx]: what about the parent animation frame? or is it too fast that we don't need to consider it? 0/5 if we cancel it here

File: components/emoji_picker/emoji_picker.test.jsx (hunk 1 above)
Comment (enahum) [reply components/emoji_picker/emoji_picker.test.jsx]: I don't think it does not hurt to have it but it doesn't do much

File: components/emoji_picker/emoji_picker.test.jsx [hunk 4]
```diff
@@ -20,4 +39,119 @@ describe('components/emoji_picker/EmojiPicker', () => {
...
+
+        // Nine categories as there is no recent caterogry
+        expect(wrapper.find(EmojiPickerCategory).length).toBe(9);
+        expect(wrapper.find(EmojiPickerCategory).find({selected: true}).length).toBe(1);
+
+        // People EmojiPickerCategory should have prop selected true
+        expect(wrapper.find(EmojiPickerCategory).find({selected: true}).find({category: 'people'}).length).toBe(1);
+
+        expect(wrapper.find(EmojiPickerItem).length).toBe(291);
```
Comment (enahum) [inline components/emoji_picker/emoji_picker.test.jsx]: I think this is can be broken rather easily by adding or removing default emoji's as in when updating the emoji set, is there another way to make it resistant?

File: components/emoji_picker/emoji_picker.test.jsx [hunk 5]
```diff
@@ -20,4 +39,119 @@ describe('components/emoji_picker/EmojiPicker', () => {
...
+        };
+
+        expect(wrapper).toMatchSnapshot();
+
+        // 10 categories as there is recent caterogry and prop selected should be true on EmojiPickerCategory for recent category
+        expect(wrapper.find(EmojiPickerCategory).length).toBe(10);
+        expect(wrapper.find(EmojiPickerCategory).find({selected: true}).length).toBe(1);
+        expect(wrapper.find(EmojiPickerCategory).find({selected: true}).find({category: 'recent'}).length).toBe(1);
+
+        expect(wrapper.find(EmojiPickerItem).length).toBe(292);
```
Comment (enahum) [inline components/emoji_picker/emoji_picker.test.jsx]: same as above

File: components/emoji_picker/emoji_picker.test.jsx [hunk 6]
```diff
@@ -20,4 +39,119 @@ describe('components/emoji_picker/EmojiPicker', () => {
...
+        jest.runAllTimers();
+        wrapper.update();
+
+        // 10 categories as there is recent caterogry and prop selected should be true on EmojiPickerCategory for recent category
+        expect(wrapper.find(EmojiPickerCategory).length).toBe(10);
+        expect(wrapper.find(EmojiPickerCategory).find({selected: true}).length).toBe(1);
+        expect(wrapper.find(EmojiPickerCategory).find({selected: true}).find({category: 'recent'}).length).toBe(1);
+
+        expect(wrapper.state('renderAllCategories')).toEqual(true);
+        expect(wrapper.find(EmojiPickerItem).length).toBe(1449);
```
Comment (enahum) [inline components/emoji_picker/emoji_picker.test.jsx]: same here

File: components/emoji_picker/emoji_picker.jsx (hunk 3 above)
Comment (sudheerDev) [reply components/emoji_picker/emoji_picker.jsx]: Ya, It is too fast so was not good enough

File: components/emoji_picker/emoji_picker.test.jsx (hunk 4 above)
Comment (sudheerDev) [reply components/emoji_picker/emoji_picker.test.jsx]: Ya, Actually we are not interested in the actual numbers here so made changes to remove these

**Comment by lindalumitchell** [comment]
LGTM; removing QA Review label.

**Comment by amyblais** [comment]
Removed from v5.16 because it didn't make it to feature complete. Please let me know if this is important for v5.16.

**Comment by esethna** [comment]
@amyblais curious why this PR didn't make it in? I was marked as reviews complete on feature complete?

**Comment by amyblais** [comment]
It hasn't been merged yet for QA. Let me know if this is important for v5.16.

**Comment by esethna** [comment]
Okay, there's nothing blocking it from being merged, I'd like to see this in v5.16 if possible.

**Comment by esethna** [comment]
Thanks, I've pinged dev to merge it

**Pull request merged**